torch/_dynamo/symbolic_convert.py

```diff
@@ -2641,33 +2641,46 @@ def GET_YIELD_FROM_ITER(self, inst):
             self.pop()
             res = BuiltinVariable(iter).call_function(self, [tos], {})
             self.push(res)
-        return self.YIELD_FROM(inst)
 
     def YIELD_FROM(self, inst):
-        while True:
-            tos = self.stack[-1].realize()
-            if isinstance(tos, ConstantVariable) and tos.value is None:
-                self.pop()
-                return
-            try:
-                val = tos.next_variable(self)
-                self.push(val)
-                # TODO(voz): Unclear if we need the push None in YIELD_VALUE?
-                self.YIELD_VALUE(inst)
-                self.pop()
-                self.push(tos)
-            except (StopIteration, exc.UserStopIteration):
-                # TODO(jansel): do we need a self.pop() here?
-                return
+        assert len(self.stack) >= 2
+        val = self.pop()
+        tos = self.stack[-1].realize()
+        if not (isinstance(val, ConstantVariable) and val.value is None):
+            # invoke send
+            # Unreachable code - if you hit this, you are implementing generator support and have
+            # lifted the `unimplemented("generator")` in frame conversion. This codepath handles
+            # subgenerator and lines up with this line in Python 3.10
+            # https://github.com/python/cpython/blob/3.10/Python/ceval.c#L2599
+            unimplemented("Unreachable sub-generator code")
+
+        try:
+            val = tos.next_variable(self)
+        except (StopIteration, exc.UserStopIteration) as ex:
+            self.pop()
+            self.push(ConstantVariable.create(ex.value))
+        else:
+            self.push(val)
+            # TODO(voz): Unclear if we need the push None in YIELD_VALUE?
+            self.YIELD_VALUE(inst)
+            assert self.instruction_pointer is not None and self.instruction_pointer > 0
+            self.instruction_pointer -= 1
 
     def SEND(self, inst):
         assert len(self.stack) >= 2
         val = self.pop()
         tos = self.stack[-1]
         if isinstance(tos, ListIteratorVariable):
             if isinstance(val, ConstantVariable) and val.value is None:
-                self.push(val)
-                self.instruction_pointer = self.indexof[inst.target]
+                try:
+                    val = tos.next_variable(self)
+                except (StopIteration, exc.UserStopIteration) as ex:
+                    if sys.version_info < (3, 12):
+                        self.pop()  # Python 3.12 uses new opcode END_SEND
+                    self.push(ConstantVariable.create(ex.value))
+                    self.instruction_pointer += inst.arg
```

Comment:
In the past there have been some changes to how bytecodes are indexed between Python versions, so I think using `inst.target` is more robust.  What is the reason for this change?

Reply:
I changed the implementation based on [Python documentation opcode `SEND(delta)`](https://docs.python.org/3.12/library/dis.html#opcode-SEND):

> `SEND(delta)`
> Equivalent to `STACK[-1] = STACK[-2].send(STACK[-1])`. Used in yield from and await statements.
> 
> If the call raises [StopIteration](https://docs.python.org/3.12/library/exceptions.html#StopIteration), pop the top value from the stack, push the exception’s value attribute, and increment the bytecode counter by `delta`.
> 
> New in version 3.11.

And the CPython implementation for opcode `SEND`:

https://github.com/python/cpython/blob/3bc0d2b851bfa38bb3614d56bd9b83b1c651ea63/Python/ceval.c#L2617

```c
if (gen_status == PYGEN_RETURN) {
    assert(retval != NULL);
    Py_DECREF(receiver);
    SET_TOP(retval);
    JUMPBY(oparg);
    DISPATCH();
}
```

Reply:
I changed it back to use `self.instruction_pointer = self.indexof[inst.target]`.